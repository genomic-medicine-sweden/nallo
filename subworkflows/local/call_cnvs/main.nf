include { ADD_FOUND_IN_TAG } from '../../../modules/local/add_found_in_tag'
include { HIFICNV          } from '../../../modules/local/pacbio/hificnv'
include { SVDB_MERGE       } from '../../../modules/nf-core/svdb/merge/main'
include { TABIX_TABIX      } from '../../../modules/nf-core/tabix/tabix/main'

workflow CALL_CNVS {

    take:
    ch_bam_bai_vcf     // channel: [ val(meta), path(bam), path(vcf) ]
    ch_fasta           // channel: [ val(meta), path(fasta) ]
    ch_expected_xy_bed // channel: [ val(meta), path(bed) ]
    ch_expected_xx_bed // channel: [ val(meta), path(bed) ]
    ch_exclude_bed     // channel: [ val(meta), path(bed) ]

    main:
    ch_versions     = Channel.empty()

    ch_bam_bai_vcf
        .map { meta, bam, bai, vcf -> [ meta, bam, bai, vcf, meta.sex ] }
        .set { ch_hificnv_in }

    // Run HiFiCNV
    HIFICNV (
        ch_hificnv_in,
        ch_fasta,
        ch_expected_xy_bed,
        ch_expected_xx_bed,
        ch_exclude_bed
    )
    ch_versions = ch_versions.mix(HIFICNV.out.versions)

    // Add FOUND_IN=hificnv to VCF
    ADD_FOUND_IN_TAG (
        HIFICNV.out.vcf.map { meta, vcf -> [ meta, vcf, [] ] },
        "hificnv"
    )
    ch_versions = ch_versions.mix(ADD_FOUND_IN_TAG.out.versions)

    ADD_FOUND_IN_TAG.out.vcf
        .map { meta, vcf -> [ [ 'id': meta.family_id ], vcf ] }
        .groupTuple()
        .set { svdb_merge_in }

    // Merge the files
    SVDB_MERGE (
        svdb_merge_in,
        [],
        true
    )
    ch_versions = ch_versions.mix(SVDB_MERGE.out.versions)

    TABIX_TABIX ( SVDB_MERGE.out.vcf )
    ch_versions = ch_versions.mix(TABIX_TABIX.out.versions)

    emit:
    sample_vcf  = ADD_FOUND_IN_TAG.out.vcf // channel: [ val(meta), path(vcf) ]
    sample_tbi  = ADD_FOUND_IN_TAG.out.tbi // channel: [ val(meta), path(tbi) ]
    family_vcf  = SVDB_MERGE.out.vcf       // channel: [ val(meta), path(vcf) ]
    family_tbi  = TABIX_TABIX.out.tbi      // channel: [ val(meta), path(tbi) ]
    versions = ch_versions                 // channel: [ versions.yml ]
}

