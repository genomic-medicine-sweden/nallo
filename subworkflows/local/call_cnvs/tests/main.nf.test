nextflow_workflow {

    name "Test Workflow CALL_CNVS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "CALL_CNVS"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.pipelines_testdata_base_path + 'reference/hg38.test.fa.gz', checkIfExists: true)
                ]
                """
            }

        }
        run("SAMTOOLS_FAIDX") {
            script "../../../../modules/nf-core/samtools/faidx/main.nf"
            process {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = [[],[]]
                """
            }
        }
        run("MINIMAP2_ALIGN") {
            script "../../../../modules/nf-core/minimap2/align/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'test', project: 'test', sex: 1 ],
                    file(params.pipelines_testdata_base_path + 'testdata/HG002_PacBio_Revio.bam', checkIfExists: true)
                ]
                input[1] = GUNZIP.out.gunzip
                input[2] = true
                input[3] = 'bai'
                input[4] = false
                input[5] = false
                """
            }
        }
        run("DEEPVARIANT") {
            script "../../../../modules/nf-core/deepvariant/main.nf"
            process {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .map { meta, bam, bai -> [ meta, bam, bai, file(params.pipelines_testdata_base_path + 'reference/test_data.bed', checkIfExists: true) ] }
                input[1] = GUNZIP.out.gunzip
                input[2] = SAMTOOLS_FAIDX.out.fai
                input[3] = [[],[]]
                input[4] = [[],[]]
                """
            }
        }

    }

    test("1 sample - [bam, bai, [] ], fasta, [], [], [], []") {

        when {
            workflow {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .map { meta, bam, bai -> [ meta, bam, bai, [] ] }
                input[1] = GUNZIP.out.gunzip
                input[2] = [[],[]]
                input[3] = [[],[]]
                input[4] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.sample_tbi.get(0).get(1).endsWith("tbi") },
                { assert workflow.out.project_tbi.get(0).get(1).endsWith("tbi") },
                { assert snapshot(
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.project_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.project_vcf.get(0).get(1)).vcf.variantsMD5,
                    workflow.out.versions,
                ).match() }
            )
        }
    }

    test("1 sample - [bam, bai, vcf ], fasta, [], [], [], []") {

        when {
            workflow {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .join(DEEPVARIANT.out.vcf)
                input[1] = GUNZIP.out.gunzip
                input[2] = [[],[]]
                input[3] = [[],[]]
                input[4] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.sample_tbi.get(0).get(1).endsWith("tbi") },
                { assert workflow.out.project_tbi.get(0).get(1).endsWith("tbi") },
                { assert snapshot(
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.project_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.project_vcf.get(0).get(1)).vcf.variantsMD5,
                    workflow.out.versions,
                ).match() }
            )
        }
    }

    test("1 sample - [bam, bai, vcf ], fasta, xy_bed, xx_bed, exclude_bed") {

        when {
            workflow {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .join(DEEPVARIANT.out.vcf)
                input[1] = GUNZIP.out.gunzip
                input[2] = [
                    [ id: 'xy_bed' ],
                    file(params.pipelines_testdata_base_path + 'reference/expected_cn.hg38.XY.bed', checkIfExists: true)
                ]
                input[3] = [
                    [ id: 'xx_bed' ],
                    file(params.pipelines_testdata_base_path + 'reference/expected_cn.hg38.XX.bed', checkIfExists: true)
                ]
                input[4] = [
                    [ id: 'exclude_bed' ],
                    file(params.pipelines_testdata_base_path + 'reference/cnv.excluded_regions.hg38.bed.gz', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.sample_tbi.get(0).get(1).endsWith("tbi") },
                { assert workflow.out.project_tbi.get(0).get(1).endsWith("tbi") },
                { assert snapshot(
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.sample_vcf.get(0).get(1)).vcf.variantsMD5,
                    path(workflow.out.project_vcf.get(0).get(1)).vcf.summary,
                    path(workflow.out.project_vcf.get(0).get(1)).vcf.variantsMD5,
                    workflow.out.versions,
                ).match() }
            )
        }
    }

    test("1 sample - [bam, bai, [] ], fasta, [], [], [], [] - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .map { meta, bam, bai -> [ meta, bam, bai, [] ] }
                input[1] = GUNZIP.out.gunzip
                input[2] = [[],[]]
                input[3] = [[],[]]
                input[4] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("1 sample - [bam, bai, vcf ], fasta, [], [], [], [] - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .join(DEEPVARIANT.out.vcf)
                input[1] = GUNZIP.out.gunzip
                input[2] = [[],[]]
                input[3] = [[],[]]
                input[4] = [[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("1 sample - [bam, bai, vcf ], fasta, xy_bed, xx_bed, exclude_bed - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = MINIMAP2_ALIGN.out.bam
                    .join(MINIMAP2_ALIGN.out.index)
                    .join(DEEPVARIANT.out.vcf)
                input[1] = GUNZIP.out.gunzip
                input[2] = [
                    [ id: 'xy_bed' ],
                    file(params.pipelines_testdata_base_path + 'reference/expected_cn.hg38.XY.bed', checkIfExists: true)
                ]
                input[3] = [
                    [ id: 'xx_bed' ],
                    file(params.pipelines_testdata_base_path + 'reference/expected_cn.hg38.XX.bed', checkIfExists: true)
                ]
                input[4] = [
                    [ id: 'exclude_bed' ],
                    file(params.pipelines_testdata_base_path + 'reference/cnv.excluded_regions.hg38.bed.gz', checkIfExists: true)
                ]
                """
            }
            then {
                assertAll(
                    { assert workflow.success },
                    { assert snapshot(workflow.out).match() }
                )
            }
        }
    }
}
