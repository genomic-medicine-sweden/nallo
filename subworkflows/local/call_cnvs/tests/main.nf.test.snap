{
    "1 sample - [bam, bai, vcf ], fasta, [], [], [], [] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                    "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                    "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                    "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
                ],
                "project_tbi": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "project_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "sample_tbi": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_vcf": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                    "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                    "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                    "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T16:07:28.10949567"
    },
    "1 sample - [bam, bai, [] ], fasta, [], [], [], [] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                    "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                    "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                    "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
                ],
                "project_tbi": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "project_vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "merged.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "sample_tbi": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz.tbi:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "sample_vcf": [
                    [
                        {
                            "id": "test",
                            "project": "test",
                            "sex": "1:md5,6507d2da8b6c8218b4ab02c6da1d59e1"
                        },
                        "test.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                    "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                    "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                    "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T15:24:06.917316228"
    },
    "1 sample - [bam, bai, vcf ], fasta, xy_bed, xx_bed, exclude_bed": {
        "content": [
            "VcfFile [chromosomes=[chrX, chr16, chr20], sampleCount=1, variantCount=36, phased=false, phasedAutodetect=false]",
            "444974b6ec3f9600ca12538628e6742c",
            "VcfFile [chromosomes=[chrX, chr16, chr20], sampleCount=1, variantCount=36, phased=false, phasedAutodetect=false]",
            "171d1b5284a06778c87f54c27ee88bea",
            [
                "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T16:07:03.510729087"
    },
    "1 sample - [bam, bai, [] ], fasta, [], [], [], []": {
        "content": [
            "VcfFile [chromosomes=[chrX, chr16, chr20], sampleCount=1, variantCount=41, phased=false, phasedAutodetect=false]",
            "576892dd42d2b2cb06542862a4beddc1",
            "VcfFile [chromosomes=[chrX, chr16, chr20], sampleCount=1, variantCount=41, phased=false, phasedAutodetect=false]",
            "4aaee123cd9279e17220409be6ed7ea3",
            [
                "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T16:00:30.337555914"
    },
    "1 sample - [bam, bai, vcf ], fasta, [], [], [], []": {
        "content": [
            "VcfFile [chromosomes=[chrX, chr16, chr20], sampleCount=1, variantCount=41, phased=false, phasedAutodetect=false]",
            "576892dd42d2b2cb06542862a4beddc1",
            "VcfFile [chromosomes=[chrX, chr16, chr20], sampleCount=1, variantCount=41, phased=false, phasedAutodetect=false]",
            "4aaee123cd9279e17220409be6ed7ea3",
            [
                "versions.yml:md5,98e5fabb16a6a19137277ce414001426",
                "versions.yml:md5,b35bbabdf1cbee70be1139199d1a2fb9",
                "versions.yml:md5,cde327ef676bc2aa6933c67299475699",
                "versions.yml:md5,fc490a5d4e1a6dce983637532901f1e9"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-16T16:04:31.143530362"
    }
}